--- a/frontend/saned.c.old	2017-06-11 14:32:19.781766759 +0200
+++ b/frontend/saned.c	2017-06-15 21:37:30.187667431 +0200
@@ -1853,7 +1853,7 @@
       {
 	SANE_Open_Reply reply;
 	SANE_Handle be_handle;
-	SANE_String name, resource;
+	SANE_String name, resource, tmp_name;
 	sem_t* sem = SEM_FAILED;
 
 	sanei_w_string (w, &name);
@@ -1927,6 +1927,19 @@
 	    memset (&reply, 0, sizeof (reply));	/* avoid leaking bits */
 
 	 		reply.status = SANE_STATUS_DEVICE_BUSY;
+			
+			/* first, notify that we want to open the device */
+			tmp_name = malloc(strlen(name)+strlen(".open")+1);
+			strcpy(tmp_name, name);
+			strcat(tmp_name, ".open");
+			sem = sem_open(tmp_name, O_CREAT, 0660, 0);
+			if(sem == SEM_FAILED)
+				DBG (DBG_ERR, "process_request: cannot open or create semaphore `%s' - %s\n",tmp_name, strerror(errno));
+			else
+				sem_post(sem);/* increment semaphore and that's all */
+			free(tmp_name);
+			sem_close(sem);
+
 			sem = sem_open(name, O_CREAT, 0660, 1);
 			if(sem == SEM_FAILED)
 				DBG (DBG_ERR, "process_request: cannot open or create semaphore `%s' - %s\n",name, strerror(errno));
